### ISSUE TITLE
Simplify zome macro: inputs and outputs

### ISSUE THREAD

**Issue by maackle** [holochain/holochain-rust#725]
Proposal to change zome function macro structure:

* change `inputs` and `outputs` to singular: `input` and `output`
* use `{}` instead of `||` to wrap input params, since on the wasm side it represents a single param with keys and values rather than distinct function parameters
* remove the named output, since it is just a return value and is never named

```rust
        // before:
            create_post: {
                inputs: |content: String, in_reply_to: Address|,
                outputs: |post: Post|,
                handler: blog::handle_check_sum
            }

        // after:
            create_post: {
                input: {content: String, in_reply_to: Address},
                output: Post,
                handler: blog::handle_check_sum
            }
```

**Comment by zippy** [comment]
I'm actually fine with this change, but here's the issue with dropping the names in general: loss of semantics.  The particular example it works because the output is Post.  But lets say the output were a  u32.  Then the semantics should have been encoded in the parameter name, i.e.  `|length: u32|` vs `|size: u32|`  Of course you "could/should" have created a `type length=u32` but this just indicates how most languages don't properly distinguish between structure and meaning.

**Comment by lucksus** [comment]
I disagree with changing `inputs` to `input` and using `{}`. The syntax for parameters of a closure in Rust is `|name: Type|` and there are several parameters. The fact that we serialize them all at once and use JSON as serialization format I would see as implementation details.

Regarding the output(s): we do have named outputs specified in our [DNA spec](https://github.com/holochain/holochain-rust/blob/develop/doc/specs/dna_manifest_specification.md) and implemented in the underlying [DNA structs](https://github.com/holochain/holochain-rust/blob/d51fe34a15c52a937890309cb1e72202abe3ec7f/core_types/src/dna/zome/capabilities.rs#L113). I would not want the macro to diverge from that. Are you implying to change the spec and DNA as well?

True, we did not (yet) implement any checks on the output. I thought we would at some point. Maybe we don't need that and just take whatever output the function returns. Shouldn't we then only assume JSON (as `JsonString`) and not have any output definition in the macro at all?
